Drop IsDeprecatedWeakRefSmartPointerException for CachedResourceClient

cdumez opened the pull request:
#### 4c281591ca8b9afa56ba5eea72528ceb6de5c505
<pre>
Drop IsDeprecatedWeakRefSmartPointerException for CachedResourceClient
<a href="https://bugs.webkit.org/show_bug.cgi?id=300775">https://bugs.webkit.org/show_bug.cgi?id=300775</a>
<a href="https://rdar.apple.com/163168617">rdar://163168617</a>

Reviewed by Geoffrey Garen.

In an earlier PR, I had tried to use CanMakeCheckedPtr but this did not
work out since quite a few clients destroy themselves when one of their
virtual function overrides get called. I am therefore making CachedResourceClient
ref-counted instead. However, since RenderObject is a CachedImageClient
and I do not want to make all RenderObjects ref-counted, I&apos;m using a
ref-counted trampoline object to forward the calls to the RenderObject.

* Source/WebCore/Modules/beacon/NavigatorBeacon.cpp:
(WebCore::NavigatorBeacon::ref const):
(WebCore::NavigatorBeacon::deref const):
(WebCore::NavigatorBeacon::from):
* Source/WebCore/Modules/beacon/NavigatorBeacon.h:
* Source/WebCore/Modules/mediasession/MediaMetadata.cpp:
(WebCore::ArtworkImageLoader::create):
(WebCore::ArtworkImageLoader::requestImageResource):
(WebCore::MediaMetadata::tryNextArtworkImage):
* Source/WebCore/Modules/mediasession/MediaMetadata.h:
* Source/WebCore/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncheckedLocalVarsCheckerExpectations:
* Source/WebCore/bindings/js/CachedModuleScriptLoader.h:
* Source/WebCore/bindings/js/CachedScriptSourceProvider.h:
* Source/WebCore/bindings/js/WebAssemblyCachedScriptSourceProvider.h:
* Source/WebCore/css/CSSFontFace.cpp:
(WebCore::CSSFontFace::appendSources):
* Source/WebCore/css/CSSFontFaceSource.cpp:
(WebCore::CSSFontFaceSource::CSSFontFaceSource):
* Source/WebCore/css/CSSFontFaceSource.h:
* Source/WebCore/css/CSSFontFaceSrcValue.cpp:
(WebCore::CSSFontFaceSrcResourceValue::fontLoadRequest):
* Source/WebCore/css/CSSFontFaceSrcValue.h:
* Source/WebCore/css/StyleRuleImport.cpp:
(WebCore::StyleRuleImport::StyleRuleImport):
* Source/WebCore/css/StyleRuleImport.h:
* Source/WebCore/dom/DataTransfer.cpp:
(WebCore::DataTransfer::~DataTransfer):
(WebCore::DataTransfer::setDragImage):
(WebCore::DragImageLoader::create):
(WebCore::DragImageLoader::imageChanged):
(WebCore::DataTransfer::moveDragState):
* Source/WebCore/dom/DataTransfer.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::fontLoadRequest):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/LoadableClassicScript.h:
* Source/WebCore/dom/LoadableSpeculationRules.h:
* Source/WebCore/dom/ProcessingInstruction.h:
* Source/WebCore/dom/ScriptExecutionContext.cpp:
(WebCore::ScriptExecutionContext::fontLoadRequest):
* Source/WebCore/dom/ScriptExecutionContext.h:
* Source/WebCore/html/HTMLImageLoader.h:
* Source/WebCore/html/HTMLLinkElement.cpp:
(WebCore::HTMLLinkElement::HTMLLinkElement):
(WebCore::HTMLLinkElement::process):
(WebCore::HTMLLinkElement::removedFromAncestor):
* Source/WebCore/html/HTMLLinkElement.h:
* Source/WebCore/html/ImageInputType.cpp:
(WebCore::ImageInputType::height const):
(WebCore::ImageInputType::width const):
* Source/WebCore/html/track/LoadableTextTrack.cpp:
(WebCore::LoadableTextTrack::scheduleLoad):
* Source/WebCore/html/track/LoadableTextTrack.h:
* Source/WebCore/inspector/InspectorAuditResourcesObject.cpp:
(WebCore::InspectorAuditResourcesObject::InspectorAuditResourcesObject):
(WebCore::InspectorAuditResourcesObject::clientForResource):
* Source/WebCore/inspector/InspectorAuditResourcesObject.h:
(WebCore::InspectorAuditResourcesObject::InspectorAuditCachedResourceClient::InspectorAuditCachedResourceClient):
(WebCore::InspectorAuditResourcesObject::InspectorAuditCachedFontClient::InspectorAuditCachedFontClient):
(WebCore::InspectorAuditResourcesObject::InspectorAuditCachedRawResourceClient::InspectorAuditCachedRawResourceClient):
(WebCore::InspectorAuditResourcesObject::InspectorAuditCachedSVGDocumentClient::InspectorAuditCachedSVGDocumentClient):
(WebCore::InspectorAuditResourcesObject::InspectorAuditCachedStyleSheetClient::InspectorAuditCachedStyleSheetClient):
* Source/WebCore/loader/ApplicationManifestLoader.cpp:
(WebCore::ApplicationManifestLoader::create):
(WebCore::ApplicationManifestLoader::startLoading):
(WebCore::ApplicationManifestLoader::notifyFinished):
* Source/WebCore/loader/ApplicationManifestLoader.h:
* Source/WebCore/loader/CrossOriginPreflightChecker.cpp:
(WebCore::CrossOriginPreflightChecker::create):
(WebCore::CrossOriginPreflightChecker::notifyFinished):
(WebCore::CrossOriginPreflightChecker::redirectReceived):
(WebCore::CrossOriginPreflightChecker::startPreflight):
(WebCore::CrossOriginPreflightChecker::protectedLoader const): Deleted.
* Source/WebCore/loader/CrossOriginPreflightChecker.h:
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::loadApplicationManifest):
(WebCore::DocumentLoader::notifyFinishedLoadingApplicationManifest):
(WebCore::DocumentLoader::didGetLoadDecisionForIcon):
* Source/WebCore/loader/DocumentLoader.h:
* Source/WebCore/loader/DocumentPrefetcher.h:
(WebCore::DocumentPrefetcher::create):
* Source/WebCore/loader/DocumentThreadableLoader.cpp:
(WebCore::DocumentThreadableLoader::makeCrossOriginAccessRequestWithPreflight):
(WebCore::DocumentThreadableLoader::clearResource):
(WebCore::DocumentThreadableLoader::preflightSuccess):
(WebCore::DocumentThreadableLoader::preflightFailure):
* Source/WebCore/loader/DocumentThreadableLoader.h:
* Source/WebCore/loader/FontLoadRequest.h:
* Source/WebCore/loader/ImageLoader.h:
* Source/WebCore/loader/LinkLoader.cpp:
(WebCore::LinkLoader::create):
(WebCore::LinkLoader::~LinkLoader):
(WebCore::LinkLoader::triggerEvents):
(WebCore::LinkLoader::triggerError):
(WebCore::createLinkPreloadResourceClient):
(WebCore::LinkLoader::preloadIfNeeded):
(WebCore::LinkLoader::cancelLoad):
(WebCore::LinkLoader::loadLink):
* Source/WebCore/loader/LinkLoader.h:
* Source/WebCore/loader/LinkLoaderClient.h:
* Source/WebCore/loader/LinkPreloadResourceClients.cpp:
(WebCore::LinkPreloadResourceClient::triggerEvents):
* Source/WebCore/loader/LinkPreloadResourceClients.h:
(WebCore::LinkPreloadDefaultResourceClient::create):
(WebCore::LinkPreloadStyleResourceClient::create):
(WebCore::LinkPreloadImageResourceClient::create):
(WebCore::LinkPreloadFontResourceClient::create):
(WebCore::LinkPreloadRawResourceClient::create):
* Source/WebCore/loader/MediaResourceLoader.h:
* Source/WebCore/loader/TextTrackLoader.cpp:
(WebCore::TextTrackLoader::create):
(WebCore::TextTrackLoader::cueLoadTimerFired):
(WebCore::TextTrackLoader::processNewCueData):
(WebCore::TextTrackLoader::corsPolicyPreventedLoad):
(WebCore::TextTrackLoader::load):
(WebCore::TextTrackLoader::newRegionsParsed):
(WebCore::TextTrackLoader::newStyleSheetsParsed):
(WebCore::TextTrackLoader::getNewCues):
(WebCore::TextTrackLoader::protectedDocument const): Deleted.
* Source/WebCore/loader/TextTrackLoader.h:
* Source/WebCore/loader/cache/CachedCSSStyleSheet.cpp:
(WebCore::CachedCSSStyleSheet::checkNotify):
* Source/WebCore/loader/cache/CachedFont.cpp:
(WebCore::CachedFont::checkNotify):
* Source/WebCore/loader/cache/CachedFontLoadRequest.h:
* Source/WebCore/loader/cache/CachedImage.cpp:
(WebCore::CachedImage::addClientWaitingForAsyncDecoding):
(WebCore::CachedImage::removeAllClientsWaitingForAsyncDecoding):
(WebCore::CachedImage::notifyObservers):
(WebCore::CachedImage::canDestroyDecodedData const):
(WebCore::CachedImage::imageFrameAvailable):
(WebCore::CachedImage::imageContentChanged):
(WebCore::CachedImage::scheduleRenderingUpdate):
(WebCore::CachedImage::allowsAnimation const):
(WebCore::CachedImage::isVisibleInViewport const):
* Source/WebCore/loader/cache/CachedImageClient.h:
* Source/WebCore/loader/cache/CachedRawResource.cpp:
(WebCore::CachedRawResource::notifyClientsDataWasReceived):
(WebCore::iterateClients):
(WebCore::CachedRawResource::responseReceived):
(WebCore::CachedRawResource::shouldCacheResponse):
(WebCore::CachedRawResource::didSendData):
(WebCore::CachedRawResource::finishedTimingForWorkerLoad):
* Source/WebCore/loader/cache/CachedRawResourceClient.h:
* Source/WebCore/loader/cache/CachedResource.cpp:
(WebCore::CachedResource::checkNotify):
(WebCore::CachedResource::switchClientsToRevalidatedResource):
* Source/WebCore/loader/cache/CachedResourceClient.h:
* Source/WebCore/loader/cache/CachedResourceLoader.cpp:
(WebCore::CachedResourceLoader::CachedResourceLoader):
* Source/WebCore/loader/cache/CachedResourceLoader.h:
(WebCore::CachedResourceLoader::keepaliveRequestTracker):
* Source/WebCore/loader/cache/CachedSVGDocumentReference.cpp:
(WebCore::CachedSVGDocumentReference::create):
* Source/WebCore/loader/cache/CachedSVGDocumentReference.h:
* Source/WebCore/loader/cache/CachedTextTrack.cpp:
(WebCore::CachedTextTrack::doUpdateBuffer):
* Source/WebCore/loader/cache/CachedXSLStyleSheet.cpp:
(WebCore::CachedXSLStyleSheet::checkNotify):
* Source/WebCore/loader/cache/KeepaliveRequestTracker.cpp:
(WebCore::KeepaliveRequestTracker::create):
* Source/WebCore/loader/cache/KeepaliveRequestTracker.h:
* Source/WebCore/loader/icon/IconLoader.cpp:
(WebCore::IconLoader::create):
(WebCore::IconLoader::startLoading):
(WebCore::IconLoader::notifyFinished):
* Source/WebCore/loader/icon/IconLoader.h:
* Source/WebCore/platform/graphics/avfoundation/objc/WebCoreAVFResourceLoader.h:
* Source/WebCore/platform/graphics/avfoundation/objc/WebCoreAVFResourceLoader.mm:
(WebCore::CachedResourceMediaLoader::create):
* Source/WebCore/rendering/BackgroundPainter.cpp:
(WebCore::BackgroundPainter::paintFillLayerImpl const):
* Source/WebCore/rendering/RenderImage.cpp:
(WebCore::RenderImage::paintReplaced):
(WebCore::RenderImage::paintIntoRect):
* Source/WebCore/rendering/RenderImageResource.cpp:
(WebCore::RenderImageResource::setCachedImage):
(WebCore::RenderImageResource::setContainerContext):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderLayerFilters.cpp:
(WebCore::RenderLayerFilters::create):
(WebCore::RenderLayerFilters::RenderLayerFilters):
(WebCore::RenderLayerFilters::notifyFinished):
(WebCore::RenderLayerFilters::updateReferenceFilterClients):
(WebCore::RenderLayerFilters::removeReferenceFilterClients):
* Source/WebCore/rendering/RenderLayerFilters.h:
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::RenderObject):
(WebCore::RenderObject::CachedImageListener::create):
(WebCore::RenderObject::CachedImageListener::CachedImageListener):
(WebCore::RenderObject::CachedImageListener::notifyFinished):
(WebCore::RenderObject::CachedImageListener::imageChanged):
(WebCore::RenderObject::CachedImageListener::allowsAnimation const):
(WebCore::RenderObject::CachedImageListener::canDestroyDecodedData const):
(WebCore::RenderObject::CachedImageListener::imageFrameAvailable):
(WebCore::RenderObject::CachedImageListener::imageVisibleInViewport const):
(WebCore::RenderObject::CachedImageListener::didRemoveCachedImageClient):
(WebCore::RenderObject::CachedImageListener::imageContentChanged):
(WebCore::RenderObject::CachedImageListener::scheduleRenderingUpdateForImage):
(WebCore::RenderObject::imageFrameAvailable):
(WebCore::RenderObject::imageChanged): Deleted.
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::notifyFinished):
(WebCore::RenderObject::allowsAnimation const):
(WebCore::RenderObject::canDestroyDecodedData const):
(WebCore::RenderObject::imageVisibleInViewport const):
(WebCore::RenderObject::didRemoveCachedImageClient):
(WebCore::RenderObject::imageContentChanged):
(WebCore::RenderObject::scheduleRenderingUpdateForImage):
(WebCore::RenderObject::protectedCachedImageClient const):
(WebCore::RenderObject::cachedImageClient const):
* Source/WebCore/rendering/style/ReferenceFilterOperation.cpp:
(WebCore::Style::ReferenceFilterOperation::loadExternalDocumentIfNeeded):
* Source/WebCore/rendering/style/ReferenceFilterOperation.h:
* Source/WebCore/rendering/style/StyleCachedImage.cpp:
(WebCore::StyleCachedImage::setContainerContextForRenderer):
(WebCore::StyleCachedImage::addClient):
(WebCore::StyleCachedImage::removeClient):
(WebCore::StyleCachedImage::hasClient const):
* Source/WebCore/rendering/style/StyleCrossfadeImage.h:
* Source/WebCore/rendering/style/StyleCursorImage.cpp:
(WebCore::StyleCursorImage::setContainerContextForRenderer):
* Source/WebCore/rendering/style/StyleFilterImage.h:
* Source/WebCore/rendering/svg/RenderSVGImage.cpp:
(WebCore::RenderSVGImage::paintIntoRect):
(WebCore::RenderSVGImage::paintForeground):
* Source/WebCore/svg/SVGFEImageElement.h:
* Source/WebCore/svg/SVGFontFaceUriElement.h:
* Source/WebCore/svg/SVGImageLoader.h:
* Source/WebCore/svg/SVGUseElement.h:
* Source/WebCore/svg/graphics/SVGImageCache.cpp:
(WebCore::SVGImageCache::findImageForRenderer const):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::loadArtworkImage):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/workers/WorkerFontLoadRequest.cpp:
(WebCore::WorkerFontLoadRequest::create):
* Source/WebCore/workers/WorkerFontLoadRequest.h:
* Source/WebCore/workers/WorkerGlobalScope.cpp:
(WebCore::WorkerGlobalScope::fontLoadRequest):
* Source/WebCore/workers/WorkerGlobalScope.h:
* Source/WebCore/xml/XSLImportRule.cpp:
(WebCore::XSLImportRule::create):
* Source/WebCore/xml/XSLImportRule.h:
* Source/WebCore/xml/XSLStyleSheet.h:
* Source/WebCore/xml/XSLStyleSheetLibxslt.cpp:
(WebCore::XSLStyleSheet::loadChildSheet):
* Source/WebKitLegacy/mac/WebView/WebHTMLView.mm:
(EmptyCachedImageClient::create):
(promisedDataClient):

Canonical link: <a href="https://commits.webkit.org/303654@main">https://commits.webkit.org/303654@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7e8d151767201763b4c94e7717b8f43f07423142

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/132684 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/5179 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/43755 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/140212 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/84711 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/7ea35a38-781f-482f-b4e3-3b42583fff4b) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/134554 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6093 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/4938 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/101442 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/68745 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/805d77c0-0fb0-45f2-80d7-0903e4f0568f) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/135630 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/4298 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/118864 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82235 "Passed tests") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/2548823d-ff9b-4814-8314-cbc4e34e4075) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/4182 "Passed tests") | [❌ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/1442 "Found 1 new API test failure: TestWebKitAPI.ObscuredContentInsets.TopOverhangColorExtensionLayer (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/83446 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/113257 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/36981 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/142867 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/4849 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/37567 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/109818 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/4931 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/4193 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/109995 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27980 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3700 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/115135 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/58324 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/4903 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/33483 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/4741 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/68354 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/4994 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/4860 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/WebCore/rendering/RenderObject.cpp:
```diff
@@ -150,14 +152,15 @@ void RenderObjectDeleter::operator() (RenderObject* renderer) const
     , m_typeFlags(node.isDocumentNode() ? (typeFlags | TypeFlag::IsAnonymous) : typeFlags)
     , m_type(type)
     , m_typeSpecificFlags(typeSpecificFlags)
+    , m_cachedImageClient(CachedImageListener::create(*this))
```
Is it possible to allocate this lazily? It feels kinda wasteful to do this allocation for every RenderObject. Probably most RenderObjects don't load images?

cdumez replied:
Good idea. I'll look into making this change now.

Note that I have perf A/B ongoing already and it is looking neutral so far.

cdumez replied:
I have made this change and fixed the few Safer CPP failures that showed up after rebasing on main.

geoffreygaren commented on Source/WebCore/rendering/RenderObject.h:
```diff
@@ -1129,6 +1139,35 @@ class RenderObject : public CachedImageClient {
 
 private:
+    // This class is to avoid making RenderObject refcounted.
+    class CachedImageListener final : public CachedImageClient, public RefCounted<CachedImageListener> {
```
You could imagine CachedImage using a type such as `RefOrCheckedPtr`, which would hold whatever kind of strong pointer the client supported, during callbacks. That's another way to honor memory safety while also honoring the requirement that RenderObject is singly owned. And then this abstraction might not be needed. 

Something to keep in mind for the future -- but I think this is a big improvement for now.

cdumez replied:
Note sure how this `RefOrCheckedPtr` would work though a single client interface like here (CachedResourceClient). How would we know if the client (subclass) is refcounted or canmakecheckedPtr?

cdumez committed:
Drop IsDeprecatedWeakRefSmartPointerException for CachedResourceClient
https://bugs.webkit.org/show_bug.cgi?id=300775
rdar://163168617

Reviewed by Geoffrey Garen.

In an earlier PR, I had tried to use CanMakeCheckedPtr but this did not
work out since quite a few clients destroy themselves when one of their
virtual function overrides get called. I am therefore making CachedResourceClient
ref-counted instead. However, since RenderObject is a CachedImageClient
and I do not want to make all RenderObjects ref-counted, I'm using a
ref-counted trampoline object to forward the calls to the RenderObject.

* Source/WebCore/Modules/beacon/NavigatorBeacon.cpp:
(WebCore::NavigatorBeacon::ref const):
(WebCore::NavigatorBeacon::deref const):
(WebCore::NavigatorBeacon::from):
* Source/WebCore/Modules/beacon/NavigatorBeacon.h:
* Source/WebCore/Modules/mediasession/MediaMetadata.cpp:
(WebCore::ArtworkImageLoader::create):
(WebCore::ArtworkImageLoader::requestImageResource):
(WebCore::MediaMetadata::tryNextArtworkImage):
* Source/WebCore/Modules/mediasession/MediaMetadata.h:
* Source/WebCore/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncheckedLocalVarsCheckerExpectations:
* Source/WebCore/bindings/js/CachedModuleScriptLoader.h:
* Source/WebCore/bindings/js/CachedScriptSourceProvider.h:
* Source/WebCore/bindings/js/WebAssemblyCachedScriptSourceProvider.h:
* Source/WebCore/css/CSSFontFace.cpp:
(WebCore::CSSFontFace::appendSources):
* Source/WebCore/css/CSSFontFaceSource.cpp:
(WebCore::CSSFontFaceSource::CSSFontFaceSource):
* Source/WebCore/css/CSSFontFaceSource.h:
* Source/WebCore/css/CSSFontFaceSrcValue.cpp:
(WebCore::CSSFontFaceSrcResourceValue::fontLoadRequest):
* Source/WebCore/css/CSSFontFaceSrcValue.h:
* Source/WebCore/css/StyleRuleImport.cpp:
(WebCore::StyleRuleImport::StyleRuleImport):
* Source/WebCore/css/StyleRuleImport.h:
* Source/WebCore/dom/DataTransfer.cpp:
(WebCore::DataTransfer::~DataTransfer):
(WebCore::DataTransfer::setDragImage):
(WebCore::DragImageLoader::create):
(WebCore::DragImageLoader::imageChanged):
(WebCore::DataTransfer::moveDragState):
* Source/WebCore/dom/DataTransfer.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::fontLoadRequest):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/LoadableClassicScript.h:
* Source/WebCore/dom/LoadableSpeculationRules.h:
* Source/WebCore/dom/ProcessingInstruction.h:
* Source/WebCore/dom/ScriptExecutionContext.cpp:
(WebCore::ScriptExecutionContext::fontLoadRequest):
* Source/WebCore/dom/ScriptExecutionContext.h:
* Source/WebCore/html/HTMLImageLoader.h:
* Source/WebCore/html/HTMLLinkElement.cpp:
(WebCore::HTMLLinkElement::HTMLLinkElement):
(WebCore::HTMLLinkElement::process):
(WebCore::HTMLLinkElement::removedFromAncestor):
* Source/WebCore/html/HTMLLinkElement.h:
* Source/WebCore/html/ImageInputType.cpp:
(WebCore::ImageInputType::height const):
(WebCore::ImageInputType::width const):
* Source/WebCore/html/track/LoadableTextTrack.cpp:
(WebCore::LoadableTextTrack::scheduleLoad):
* Source/WebCore/html/track/LoadableTextTrack.h:
* Source/WebCore/inspector/InspectorAuditResourcesObject.cpp:
(WebCore::InspectorAuditResourcesObject::InspectorAuditResourcesObject):
(WebCore::InspectorAuditResourcesObject::clientForResource):
* Source/WebCore/inspector/InspectorAuditResourcesObject.h:
(WebCore::InspectorAuditResourcesObject::InspectorAuditCachedResourceClient::InspectorAuditCachedResourceClient):
(WebCore::InspectorAuditResourcesObject::InspectorAuditCachedFontClient::InspectorAuditCachedFontClient):
(WebCore::InspectorAuditResourcesObject::InspectorAuditCachedRawResourceClient::InspectorAuditCachedRawResourceClient):
(WebCore::InspectorAuditResourcesObject::InspectorAuditCachedSVGDocumentClient::InspectorAuditCachedSVGDocumentClient):
(WebCore::InspectorAuditResourcesObject::InspectorAuditCachedStyleSheetClient::InspectorAuditCachedStyleSheetClient):
* Source/WebCore/loader/ApplicationManifestLoader.cpp:
(WebCore::ApplicationManifestLoader::create):
(WebCore::ApplicationManifestLoader::startLoading):
(WebCore::ApplicationManifestLoader::notifyFinished):
* Source/WebCore/loader/ApplicationManifestLoader.h:
* Source/WebCore/loader/CrossOriginPreflightChecker.cpp:
(WebCore::CrossOriginPreflightChecker::create):
(WebCore::CrossOriginPreflightChecker::notifyFinished):
(WebCore::CrossOriginPreflightChecker::redirectReceived):
(WebCore::CrossOriginPreflightChecker::startPreflight):
(WebCore::CrossOriginPreflightChecker::protectedLoader const): Deleted.
* Source/WebCore/loader/CrossOriginPreflightChecker.h:
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::loadApplicationManifest):
(WebCore::DocumentLoader::notifyFinishedLoadingApplicationManifest):
(WebCore::DocumentLoader::didGetLoadDecisionForIcon):
* Source/WebCore/loader/DocumentLoader.h:
* Source/WebCore/loader/DocumentPrefetcher.h:
(WebCore::DocumentPrefetcher::create):
* Source/WebCore/loader/DocumentThreadableLoader.cpp:
(WebCore::DocumentThreadableLoader::makeCrossOriginAccessRequestWithPreflight):
(WebCore::DocumentThreadableLoader::clearResource):
(WebCore::DocumentThreadableLoader::preflightSuccess):
(WebCore::DocumentThreadableLoader::preflightFailure):
* Source/WebCore/loader/DocumentThreadableLoader.h:
* Source/WebCore/loader/FontLoadRequest.h:
* Source/WebCore/loader/ImageLoader.h:
* Source/WebCore/loader/LinkLoader.cpp:
(WebCore::LinkLoader::create):
(WebCore::LinkLoader::~LinkLoader):
(WebCore::LinkLoader::triggerEvents):
(WebCore::LinkLoader::triggerError):
(WebCore::createLinkPreloadResourceClient):
(WebCore::LinkLoader::preloadIfNeeded):
(WebCore::LinkLoader::cancelLoad):
(WebCore::LinkLoader::loadLink):
* Source/WebCore/loader/LinkLoader.h:
* Source/WebCore/loader/LinkLoaderClient.h:
* Source/WebCore/loader/LinkPreloadResourceClients.cpp:
(WebCore::LinkPreloadResourceClient::triggerEvents):
* Source/WebCore/loader/LinkPreloadResourceClients.h:
(WebCore::LinkPreloadDefaultResourceClient::create):
(WebCore::LinkPreloadStyleResourceClient::create):
(WebCore::LinkPreloadImageResourceClient::create):
(WebCore::LinkPreloadFontResourceClient::create):
(WebCore::LinkPreloadRawResourceClient::create):
* Source/WebCore/loader/MediaResourceLoader.h:
* Source/WebCore/loader/TextTrackLoader.cpp:
(WebCore::TextTrackLoader::create):
(WebCore::TextTrackLoader::cueLoadTimerFired):
(WebCore::TextTrackLoader::processNewCueData):
(WebCore::TextTrackLoader::corsPolicyPreventedLoad):
(WebCore::TextTrackLoader::load):
(WebCore::TextTrackLoader::newRegionsParsed):
(WebCore::TextTrackLoader::newStyleSheetsParsed):
(WebCore::TextTrackLoader::getNewCues):
(WebCore::TextTrackLoader::protectedDocument const): Deleted.
* Source/WebCore/loader/TextTrackLoader.h:
* Source/WebCore/loader/cache/CachedCSSStyleSheet.cpp:
(WebCore::CachedCSSStyleSheet::checkNotify):
* Source/WebCore/loader/cache/CachedFont.cpp:
(WebCore::CachedFont::checkNotify):
* Source/WebCore/loader/cache/CachedFontLoadRequest.h:
* Source/WebCore/loader/cache/CachedImage.cpp:
(WebCore::CachedImage::addClientWaitingForAsyncDecoding):
(WebCore::CachedImage::removeAllClientsWaitingForAsyncDecoding):
(WebCore::CachedImage::notifyObservers):
(WebCore::CachedImage::canDestroyDecodedData const):
(WebCore::CachedImage::imageFrameAvailable):
(WebCore::CachedImage::imageContentChanged):
(WebCore::CachedImage::scheduleRenderingUpdate):
(WebCore::CachedImage::allowsAnimation const):
(WebCore::CachedImage::isVisibleInViewport const):
* Source/WebCore/loader/cache/CachedImageClient.h:
* Source/WebCore/loader/cache/CachedRawResource.cpp:
(WebCore::CachedRawResource::notifyClientsDataWasReceived):
(WebCore::iterateClients):
(WebCore::CachedRawResource::responseReceived):
(WebCore::CachedRawResource::shouldCacheResponse):
(WebCore::CachedRawResource::didSendData):
(WebCore::CachedRawResource::finishedTimingForWorkerLoad):
* Source/WebCore/loader/cache/CachedRawResourceClient.h:
* Source/WebCore/loader/cache/CachedResource.cpp:
(WebCore::CachedResource::checkNotify):
(WebCore::CachedResource::switchClientsToRevalidatedResource):
* Source/WebCore/loader/cache/CachedResourceClient.h:
* Source/WebCore/loader/cache/CachedResourceLoader.cpp:
(WebCore::CachedResourceLoader::CachedResourceLoader):
* Source/WebCore/loader/cache/CachedResourceLoader.h:
(WebCore::CachedResourceLoader::keepaliveRequestTracker):
* Source/WebCore/loader/cache/CachedSVGDocumentReference.cpp:
(WebCore::CachedSVGDocumentReference::create):
* Source/WebCore/loader/cache/CachedSVGDocumentReference.h:
* Source/WebCore/loader/cache/CachedTextTrack.cpp:
(WebCore::CachedTextTrack::doUpdateBuffer):
* Source/WebCore/loader/cache/CachedXSLStyleSheet.cpp:
(WebCore::CachedXSLStyleSheet::checkNotify):
* Source/WebCore/loader/cache/KeepaliveRequestTracker.cpp:
(WebCore::KeepaliveRequestTracker::create):
* Source/WebCore/loader/cache/KeepaliveRequestTracker.h:
* Source/WebCore/loader/icon/IconLoader.cpp:
(WebCore::IconLoader::create):
(WebCore::IconLoader::startLoading):
(WebCore::IconLoader::notifyFinished):
* Source/WebCore/loader/icon/IconLoader.h:
* Source/WebCore/platform/graphics/avfoundation/objc/WebCoreAVFResourceLoader.h:
* Source/WebCore/platform/graphics/avfoundation/objc/WebCoreAVFResourceLoader.mm:
(WebCore::CachedResourceMediaLoader::create):
* Source/WebCore/rendering/BackgroundPainter.cpp:
(WebCore::BackgroundPainter::paintFillLayerImpl const):
* Source/WebCore/rendering/RenderImage.cpp:
(WebCore::RenderImage::paintReplaced):
(WebCore::RenderImage::paintIntoRect):
* Source/WebCore/rendering/RenderImageResource.cpp:
(WebCore::RenderImageResource::setCachedImage):
(WebCore::RenderImageResource::setContainerContext):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderLayerFilters.cpp:
(WebCore::RenderLayerFilters::create):
(WebCore::RenderLayerFilters::RenderLayerFilters):
(WebCore::RenderLayerFilters::notifyFinished):
(WebCore::RenderLayerFilters::updateReferenceFilterClients):
(WebCore::RenderLayerFilters::removeReferenceFilterClients):
* Source/WebCore/rendering/RenderLayerFilters.h:
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::RenderObject):
(WebCore::RenderObject::CachedImageListener::create):
(WebCore::RenderObject::CachedImageListener::CachedImageListener):
(WebCore::RenderObject::CachedImageListener::notifyFinished):
(WebCore::RenderObject::CachedImageListener::imageChanged):
(WebCore::RenderObject::CachedImageListener::allowsAnimation const):
(WebCore::RenderObject::CachedImageListener::canDestroyDecodedData const):
(WebCore::RenderObject::CachedImageListener::imageFrameAvailable):
(WebCore::RenderObject::CachedImageListener::imageVisibleInViewport const):
(WebCore::RenderObject::CachedImageListener::didRemoveCachedImageClient):
(WebCore::RenderObject::CachedImageListener::imageContentChanged):
(WebCore::RenderObject::CachedImageListener::scheduleRenderingUpdateForImage):
(WebCore::RenderObject::imageFrameAvailable):
(WebCore::RenderObject::imageChanged): Deleted.
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::notifyFinished):
(WebCore::RenderObject::allowsAnimation const):
(WebCore::RenderObject::canDestroyDecodedData const):
(WebCore::RenderObject::imageVisibleInViewport const):
(WebCore::RenderObject::didRemoveCachedImageClient):
(WebCore::RenderObject::imageContentChanged):
(WebCore::RenderObject::scheduleRenderingUpdateForImage):
(WebCore::RenderObject::protectedCachedImageClient const):
(WebCore::RenderObject::cachedImageClient const):
* Source/WebCore/rendering/style/ReferenceFilterOperation.cpp:
(WebCore::Style::ReferenceFilterOperation::loadExternalDocumentIfNeeded):
* Source/WebCore/rendering/style/ReferenceFilterOperation.h:
* Source/WebCore/rendering/style/StyleCachedImage.cpp:
(WebCore::StyleCachedImage::setContainerContextForRenderer):
(WebCore::StyleCachedImage::addClient):
(WebCore::StyleCachedImage::removeClient):
(WebCore::StyleCachedImage::hasClient const):
* Source/WebCore/rendering/style/StyleCrossfadeImage.h:
* Source/WebCore/rendering/style/StyleCursorImage.cpp:
(WebCore::StyleCursorImage::setContainerContextForRenderer):
* Source/WebCore/rendering/style/StyleFilterImage.h:
* Source/WebCore/rendering/svg/RenderSVGImage.cpp:
(WebCore::RenderSVGImage::paintIntoRect):
(WebCore::RenderSVGImage::paintForeground):
* Source/WebCore/svg/SVGFEImageElement.h:
* Source/WebCore/svg/SVGFontFaceUriElement.h:
* Source/WebCore/svg/SVGImageLoader.h:
* Source/WebCore/svg/SVGUseElement.h:
* Source/WebCore/svg/graphics/SVGImageCache.cpp:
(WebCore::SVGImageCache::findImageForRenderer const):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::loadArtworkImage):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/workers/WorkerFontLoadRequest.cpp:
(WebCore::WorkerFontLoadRequest::create):
* Source/WebCore/workers/WorkerFontLoadRequest.h:
* Source/WebCore/workers/WorkerGlobalScope.cpp:
(WebCore::WorkerGlobalScope::fontLoadRequest):
* Source/WebCore/workers/WorkerGlobalScope.h:
* Source/WebCore/xml/XSLImportRule.cpp:
(WebCore::XSLImportRule::create):
* Source/WebCore/xml/XSLImportRule.h:
* Source/WebCore/xml/XSLStyleSheet.h:
* Source/WebCore/xml/XSLStyleSheetLibxslt.cpp:
(WebCore::XSLStyleSheet::loadChildSheet):
* Source/WebKitLegacy/mac/WebView/WebHTMLView.mm:
(EmptyCachedImageClient::create):
(promisedDataClient):

Canonical link: https://commits.webkit.org/303654@main
